File: lib/RubyGL/Native/window.rb
require 'ffi'

module RubyGL::Native
	attach_function :initWindow, [], :int
	attach_function :quitWindow, [], :void
	attach_function :createWindow, [:string, :int, :int, :int, :int, :uint32], :pointer
	attach_function :destroyWindow, [:pointer], :void
	attach_function :hideWindow, [:pointer], :void
	attach_function :showWindow, [:pointer], :void
	attach_function :setWindowBrightness, [:pointer, :float], :int
	attach_function :showSimpleMessageBox, [:uint32, :string, :string, :pointer], :int
	
	attach_variable :FULLSCREEN, :int
	attach_variable :OPENGL, :int
	attach_variable :SHOWN, :int
	attach_variable :HIDDEN, :int
	attach_variable :BORDERLESS, :int
	attach_variable :RESIZABLE, :int
	attach_variable :MINIMIZED, :int
	attach_variable :MAXIMIZED, :int
	attach_variable :INPUT_FOCUS, :int
	attach_variable :MOUSE_FOCUS, :int
	attach_variable :FULLSCREEN_DESKTOP, :int
	attach_variable :ALLOW_HIGHDPI, :int
end

File: lib/RubyGL/Native/src/Window.c
#include "Window.h"

const SDL_WindowFlags FULLSCREEN = SDL_WINDOW_FULLSCREEN,
    OPENGL = SDL_WINDOW_OPENGL,
    SHOWN = SDL_WINDOW_SHOWN, 
    HIDDEN = SDL_WINDOW_HIDDEN,
    BORDERLESS = SDL_WINDOW_BORDERLESS, 
    RESIZABLE = SDL_WINDOW_RESIZABLE, 
    MINIMIZED = SDL_WINDOW_MINIMIZED,
    MAXIMIZED = SDL_WINDOW_MAXIMIZED, 
    INPUT_GRABBED = SDL_WINDOW_INPUT_GRABBED, 
    INPUT_FOCUS = SDL_WINDOW_INPUT_FOCUS,
    MOUSE_FOCUS = SDL_WINDOW_MOUSE_FOCUS, 
    FULLSCREEN_DESKTOP = SDL_WINDOW_FULLSCREEN_DESKTOP, 
    FOREIGN = SDL_WINDOW_FOREIGN,
    ALLOW_HIGHDPI = SDL_WINDOW_ALLOW_HIGHDPI;

int initWindow()
{
    return SDL_InitSubSystem(SDL_INIT_VIDEO);
}

void quitWindow()
{
    SDL_QuitSubSystem(SDL_INIT_VIDEO);
}
    
SDL_Window* createWindow(const char* title, int x, int y, int w, int h, 
                         Uint32 flags)
{
    return SDL_CreateWindow(title, x, y, w, h, flags);
}

void destroyWindow(SDL_Window* window)
{
    SDL_DestroyWindow(window);
}
                         
void hideWindow(SDL_Window* window)
{
    SDL_HideWindow(window);
}

void showWindow(SDL_Window* window)
{
    SDL_ShowWindow(window);
}

int setWindowBrightness(SDL_Window* window, float brightness)
{
    return SDL_SetWindowBrightness(window, brightness);
}

int showSimpleMessageBox(Uint32 flags, const char* title, const char* message,
                         SDL_Window* window)
{
    return SDL_ShowSimpleMessageBox(flags, title, message, window);
}
                      
/* void setGLAttributes(int majorVersion, int minorVersion, int depthSize, 
                     int doubleBuffer)
{
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MAJOR_VERSION, majorVersion);
    SDL_GL_SetAttribute(SDL_GL_CONTEXT_MINOR_VERSION, minorVersion);

    SDL_GL_SetAttribute(SDL_GL_DEPTH_SIZE, depthSize);
    SDL_GL_SetAttribute(SDL_GL_DOUBLEBUFFER, doubleBuffer);
}

SDL_Window* createGLWindow(int majorVersion, int minorVersion, int xPosition, 
                           int yPosition, int width, int height, 
                           const char* title, Uint32 windowFlags)
{
    setGLAttributes(majorVersion, minorVersion, 24, 1);

    SDL_Window* mainWindow = SDL_CreateWindow(
        title,
        xPosition,
        yPosition,
        width,
        height,
        windowFlags | SDL_WINDOW_OPENGL
    );

    return mainWindow;
}

SDL_GLContext createGLContext(SDL_Window* window)
{
    SDL_GLContext mainContext = SDL_GL_CreateContext(window);


    SDL_GL_MakeCurrent(window, mainContext);
    SDL_GL_SetSwapInterval(1);

    return mainContext;
}
 */